## Exploratory Data Analysis on the NYC AIRBNB DATASET

In [48]:
import wandb
import pandas as pd
import pandas_profiling

###  load in the airbnb dataset

In [49]:
#intialize w&b
run = wandb.init(project="nyc_airbnb", group="eda", save_code=True)
#download file from w&b
local_path = wandb.use_artifact("sample.csv:latest").file()
#load data 
df = pd.read_csv(local_path)

## explore the data

In [50]:
profile = pandas_profiling.ProfileReport(df)
profile.to_widgets()

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

C:\Users\ndrs\anaconda3\lib\site-packages\multimethod\__init__.py:315: FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  return func(*args, **kwargs)


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render widgets:   0%|          | 0/1 [00:00<?, ?it/s]

IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)

IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)

IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [61]:
# test for additional context in name column
display('beach instance',df[df['name'].str.contains("beach", na=False)].head(10))
display('luxury instance',df[df['name'].str.contains("luxury", na=False)].head(10))
display('sq instance',df[df['name'].str.contains("sq|sqft", na=False)].head(10))

'beach instance'

id                                               name   host_id  \
1010     409666  1 bdr apt, sunny & artsy, 4 min walk to the beach    405225   
1629     741354                 Simple and cozy place on the beach   2577615   
8373    6437262     Cozy private room, close to the subway & beach  33605594   
8449    6492185                A beautiful and friendly beach area  33929905   
9207    7043746                 2 block walk from the beach in NYC  36935244   
11238   8713401                 Walk to beach...Brighton Beach 2BD   1613244   
13689  10267693  Cute beach bungalow by the beach, 80min Manhattan  52793525   
14868  11790749                Big, bright apartment by the beach.  46798824   
15185  12130236                     Cozy studio steps to the beach  65096495   
15272  12227645            The beach - it's just across the street  54129736   

      host_name neighbourhood_group    neighbourhood  latitude  longitude  \
1010       Olga            Brooklyn   Brighton Beach  40.57810  -73.95455   
1629       Ella            Brooklyn   Brighton Beach  40.57728  -73.95362   
8373      Misha            Brooklyn   Sheepshead Bay  40.59225  -73.95026   
8449       Oleg            Brooklyn  Manhattan Beach  40.57872  -73.95251   
9207   Vladimir              Queens          Arverne  40.59270  -73.78990   
11238     Ariel            Brooklyn   Brighton Beach  40.58176  -73.96264   
13689      J.C.              Queens     Far Rockaway  40.59607  -73.75911   
14868      Matt            Brooklyn   Brighton Beach  40.57844  -73.95822   
15185      Eric              Queens     Belle Harbor  40.57941  -73.84844   
15272     Donna              Queens   Rockaway Beach  40.58341  -73.81650   

             room_type  price  minimum_nights  number_of_reviews last_review  \
1010   Entire home/apt     75               2                  8  2016-09-18   
1629   Entire home/apt     73              30                 21  2017-04-18   
8373      Private room     75               1                  8  2019-05-15   
8449   Entire home/apt    145               4                 59  2019-06-17   
9207   Entire home/apt    250               2                 89  2019-05-26   
11238  Entire home/apt    110              30                  7  2018-08-08   
13689  Entire home/apt    100               6                  9  2019-05-10   
14868  Entire home/apt    199               1                  0         NaN   
15185     Private room    225               1                 35  2019-06-30   
15272     Private room     88               2                 32  2017-09-04   

       reviews_per_month  calculated_host_listings_count  availability_365  
1010                0.10                               1                 2  
1629                0.26                               1               251  
8373                0.26                               1               311  
8449                1.24                               1               157  
9207                1.82                               1               300  
11238               0.19                               9               284  
13689               0.22                               1                 0  
14868                NaN                               1                 0  
15185               0.92                               1               364  
15272               0.84                               1                 0

'luxury instance'

id                                               name   host_id  \
108     25696                   1,800 sq foot in luxury building    107628   
201     53477                                3 floors of luxury!    204539   
755    271950  Huge luxury 1BR apt near Central Park South(4 ...   1362808   
916    339456   Bright, new luxury apartment in doorman building   1723485   
1028   424192                 luxury 1 BedRoom Apt Hells Kitchen   2108853   
1754   794156                       Lincoln Center  luxury condo     58366   
2263  1077179            Large luxury 1 bedroom loft in Gramercy   5927702   
2997  1746726        Chelsea NYC luxury 2 bed/2 bath, garden apt   4051126   
3170  1890851                Stylish,Sunny, spacious luxury loft   6321587   
3314  1979848                         1 BR  in a luxury apt bldg  10217559   

           host_name neighbourhood_group       neighbourhood  latitude  \
108             Dena           Manhattan              Harlem  40.80920   
201             Mark              Queens      Middle Village  40.71546   
755              Max           Manhattan     Upper East Side  40.76373   
916             Mary           Manhattan     Upper West Side  40.79526   
1028           Jasen           Manhattan      Hell's Kitchen  40.75590   
1754           Linna           Manhattan     Upper West Side  40.77445   
2263          Shayne           Manhattan            Gramercy  40.73775   
2997            Kris           Manhattan             Chelsea  40.74188   
3170             Bud            Brooklyn  Bedford-Stuyvesant  40.68166   
3314  Mary Elizabeth            Brooklyn          Park Slope  40.67122   

      longitude        room_type  price  minimum_nights  number_of_reviews  \
108   -73.94421     Private room    100               2                170   
201   -73.87854  Entire home/apt    265               7                 38   
755   -73.96897  Entire home/apt    300               3                  5   
916   -73.96550  Entire home/apt    387               1                 23   
1028  -73.99469  Entire home/apt    325              30                 18   
1754  -73.98496  Entire home/apt    210              30                 48   
2263  -73.98324  Entire home/apt    250              30                 17   
2997  -73.99729  Entire home/apt    555               3                111   
3170  -73.94325  Entire home/apt    149               2                 20   
3314  -73.98618  Entire home/apt    175               1                  0   

     last_review  reviews_per_month  calculated_host_listings_count  \
108   2019-06-23               1.61                               1   
201   2019-04-27               0.38                               5   
755   2018-12-31               0.08                               1   
916   2018-12-31               0.63                               1   
1028  2016-05-21               0.21                               1   
1754  2019-01-03               0.60                               1   
2263  2019-01-31               0.39                               2   
2997  2019-07-03               1.63                               1   
3170  2019-06-09               0.31                               1   
3314         NaN                NaN                               1   

      availability_365  
108                346  
201                365  
755                 27  
916                  0  
1028               364  
1754                36  
2263               327  
2997               191  
3170               355  
3314                 0

'sq instance'

id                                               name  host_id  \
100     22918   loft bed - near transportation-15min to times sq    32294   
108     25696                   1,800 sq foot in luxury building   107628   
711    261781                  1500+ sq ft 2BR West Village Loft  1359519   
1161   494937        The SoHo Loft - Huge Penthouse - 1,200 sqft  1527535   
1203   512775                The Cottage / 1500 sqft. of Privacy  2396295   
1305   568684                800sqft apartment with huge terrace  2798644   
1341   587740      Designer open space in TriBeCa Soho 1000sq ft  2899508   
1518   688652                   Huge Sunny Modern Apt. (1k+sqft)    62855   
2135   981494                 Large and sunny one BR (600 sq.ft)  5207582   
2199  1027808  East Village Loft - 1300 sq/ft home - 2 Bedrooms!  5655889   

              host_name neighbourhood_group        neighbourhood  latitude  \
100     Ssameer Or Trip           Manhattan               Harlem  40.82279   
108                Dena           Manhattan               Harlem  40.80920   
711               Nadya           Manhattan         West Village  40.73378   
1161         Jean-Marie           Manhattan                 SoHo  40.72162   
1203            Richard              Queens     Long Island City  40.75296   
1305         Alessandra            Brooklyn             Bushwick  40.70202   
1341     Tom & Jennifer           Manhattan                 SoHo  40.72372   
1518             Andrew            Brooklyn         Clinton Hill  40.69445   
2135             Olivia           Manhattan  Morningside Heights  40.81062   
2199  East Village Loft           Manhattan         East Village  40.72351   

      longitude        room_type  price  minimum_nights  number_of_reviews  \
100   -73.95139     Private room     60               3                 11   
108   -73.94421     Private room    100               2                170   
711   -74.00429  Entire home/apt    375               7                  8   
1161  -74.00414  Entire home/apt    499               2                 63   
1203  -73.93716  Entire home/apt    350               2                182   
1305  -73.92402  Entire home/apt    115             370                  6   
1341  -74.00277  Entire home/apt    199              60                 11   
1518  -73.96401  Entire home/apt    249               4                 19   
2135  -73.96015  Entire home/apt    100               1                  0   
2199  -73.98283  Entire home/apt    499               5                 47   

     last_review  reviews_per_month  calculated_host_listings_count  \
100   2019-01-03               0.87                               3   
108   2019-06-23               1.61                               1   
711   2017-12-30               0.12                               1   
1161  2019-07-07               0.76                               1   
1203  2019-06-11               2.20                               1   
1305  2018-04-15               0.09                               1   
1341  2019-06-21               0.13                               1   
1518  2019-06-07               0.23                               1   
2135         NaN                NaN                               1   
2199  2019-06-08               0.63                               1   

      availability_365  
100                219  
108                346  
711                  0  
1161                26  
1203               272  
1305               365  
1341               281  
1518               362  
2135                 0  
2199               341

In [33]:
#explore number of reviews more
df[df['number_of_reviews']==0]

id                                               name   host_id  \
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
19         7750                  Huge 2 BR Upper East  Cental Park     17985   
26         8700  Magnifique Suite au N de Manhattan - vue Cloitres     26394   
36        11452                        Clean and Quiet in Brooklyn      7355   
38        11943                          Country space in the city     45445   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

             host_name neighbourhood_group       neighbourhood  latitude  \
2            Elisabeth           Manhattan              Harlem  40.80902   
19                Sing           Manhattan         East Harlem  40.79685   
26     Claude & Sophie           Manhattan              Inwood  40.86754   
36                  Vt            Brooklyn  Bedford-Stuyvesant  40.68876   
38             Harriet            Brooklyn            Flatbush  40.63702   
...                ...                 ...                 ...       ...   
48890          Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891          Marisol            Brooklyn            Bushwick  40.70184   
48892    Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893              Taz           Manhattan      Hell's Kitchen  40.75751   
48894       Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
2      -73.94190     Private room    150               3                  0   
19     -73.94872  Entire home/apt    190               7                  0   
26     -73.92639     Private room     80               4                  0   
36     -73.94312     Private room     35              60                  0   
38     -73.96327     Private room    150               1                  0   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
2             NaN                NaN                               1   
19            NaN                NaN                               2   
26            NaN                NaN                               1   
36            NaN                NaN                               1   
38            NaN                NaN                               1   
...           ...                ...                             ...   
48890         NaN                NaN                               2   
48891         NaN                NaN                               2   
48892         NaN                NaN                               1   
48893         NaN                NaN                               6   
48894         NaN                NaN                               1   

       availability_365  
2                   365  
19                  249  
26                    0  
36                  365  
38                  365  
...                 ...  
48890                 9  
48891                36  
48892                27  
48893                 2  
48894                23  

[10052 rows 

In [14]:
#explore minimum nights for outliers 
df['minimum_nights'].value_counts()[:20]

C:\Users\ndrs\AppData\Local\Temp\ipykernel_22984\1983701618.py:1: FutureWarning: The behavior of `series[i:j]` with an integer-dtype index is deprecated. In a future version, this will be treated as *label-based* indexing, consistent with e.g. `series[i]` lookups. To retain the old behavior, use `series.iloc[i:j]`. To get the future behavior, use `series.loc[i:j]`.
  df['minimum_nights'].value_counts()[:20]


1     12148
2     11199
3      7506
30     3534
4      3106
5      2854
7      1975
6       694
14      543
10      464
29      328
15      272
20      216
31      191
28      174
21      133
8       129
60      101
90       95
12       88
Name: minimum_nights, dtype: int64

### Findings:
 - price has quite a few outliers that makes its distribution very odd.
 - There are listings that are "Free" or zero. This must be a wrong entry as integers/floats don't usually allow nulls and mustve been replaced with zeroes. The upper limit however may need to be tweaked. There are instances where price exceeds 1000 but does skew our dataset by quite a bit. This value may need to be tweaked/adjusted accordingly
 - there are a lot of listings which has 0 number_of_reviews(10052 rows). This may need to be explored further during training phase as this column may negatively affect our models bias.
 -  longitude, latitude and price has correlation which makes sense from a logical standpoint
 -  minimum_nights appears to have some potential outliers. There are plenty of locations which require more than >90 days and one location in particular requires 1250 day requirement.
 - the column "name" appears to have additional contextual information that may tie/help explain their listing price further which could be extracted. Eg. there are places that are listed as apartments/house. Some locations mention they are close to the beach. some locations mention how big the apartment is.


### simple outlier evaluation and removal
 - try different values (this is not final)
 - based on an educated guess anything lower than 10  and anything higher than at least 350 should be too extreme. the higher bound can be tweaked/experimented with.

In [8]:
# Drop outliers
min_price = 10
max_price = 350
idx = df['price'].between(min_price, max_price)
df = df[idx].copy()
# Convert last_review to datetime
df['last_review'] = pd.to_datetime(df['last_review'])

In [62]:
#explore/count nulls after cleaning
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

In [6]:
#force close w&b instance
run.finish()